src/lib/adapters/WebDav.js

```diff
@@ -0,0 +1,400 @@
+        }
+
+        if (response.status === 401) {
+            throw new Error('Couldn\'t authenticate for removing bookmarks from the server.');
+        }
+
+        if (response.status !== 200) {
+            return {
+                'status' : response.status,
+                'db' : new Map ()
+            };
+        }
+
+        let bookmark_array = await response.json();
+        let server_db = new Map ();
+
+        bookmark_array.forEach ( (bm) => {
+            server_db.set(bm.id, {
+                id: bm.id
+                , url: bm.url
+                , title: bm.title
+                , path: bm.path
+            });
+        });
+
+console.log ("response out");
+console.log (server_db);
+
+        return {
+            'status' : response.status,
+            'db' : server_db
+        };
+    }
+
+    async syncStart () {
+        console.log ("syncStart: started");
+        try {
+            let resp = await this.pullFromServer ();
+            this.db = resp.db;
+
+console.log ("resp");
+console.log (resp);
+
+            if (resp.status !== 200)
+            {
+                if (resp.status !== 404)
+                {
+                    throw new Error('Failed to fetch bookmarks :' + resp.status + ":");
+                }
+            }
+        } catch (e) {
+            console.log ("caught error");
+            console.log (e);
+
+            this.db = new Map ();
+        }
+
+        console.log ("syncStart: completed");
+    }
+
+    async pullBookmarks () {
+        console.log('Fetching bookmarks', this.server)
+
+        let server_db;
+        try {
+            let resp = await this.pullFromServer ();
+            server_db = resp.db;
+
+            if (resp.status !== 200)
+            {
+                if (resp.status !== 404)
+                {
+                    throw new Error('Failed to fetch bookmarks :' + resp.status + ":");
+                }
+            }
+        } catch (e) {
+            console.log ("caught error");
+            console.log (e);
+
+            server_db = new Map ();
+        }
+
+        let myBookmarks = Array.from(server_db.values())
+            .map(bm => {
+                    return new Bookmark(bm.id, null, bm.url, bm.title, bm.path)
+                    })
+
+        console.log('Received bookmarks from server')
+        console.log (myBookmarks);
+
+        return myBookmarks
+    }
+
+    async getBookmark (id, autoupdate) {
+        console.log('Fetching single bookmark', this.server);
+        let bm = this.db.get(id);
+        if (!bm) {
+            throw new Error('Failed to fetch bookmark');
+        }
+        let bookmark = new Bookmark(bm.id, null, bm.url, bm.title, bm.path);
+        return bookmark;
+    }
+
+    async createBookmark (bm) {
+        console.log('Create single bookmark', bm, this.server)
+            if (!~['https:', 'http:', 'ftp:'].indexOf(url.parse(bm.url).protocol)) {
```

Comment:
You shouldn't need this in this adapter, as it only applies to nextcloud bookmarks, which can only handle these protocols.

Reply:
If it will not be called, I will remove it.

Reply:
But I believe it gets called heavily in the case where this is the first time the extension syncs.

Reply:
Oh, not the whole method. Just the if-statement filtering out any protocols other than http(s) / ftp. I put this in place in the Nextcloud adapter, because nextcloud bookmarks doesn't like anything else, but in a custom file, we don't need that restriction.

Reply:
I will fix that.